### PR TITLE
BUG: Inconsistent behavior in Index.difference

### LINKED ISSUE #35217: BUG: Inconsistent behavior in Index.difference
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

`Index.difference` seems to be treating float vs object / int vs object differently.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
float_index = pd.Index([1.0, 2, 3])
string_index = pd.Index(['1','2','3'])
integer_index = pd.Index([1, 2, 3])
print(float_index.difference(string_index))
print(string_index.difference(float_index))

print(integer_index.difference(string_index))
print(string_index.difference(integer_index))


Float64Index([], dtype='float64')
Index([], dtype='object')
Int64Index([1, 2, 3], dtype='int64')
Index(['1', '2', '3'], dtype='object')
```

#### Problem description
Index.difference should probably not type-cast and compare the elements to honor consistent behavior with other types where they are not being type-casted.

#### Expected Output

```python
Float64Index([1.0, 2.0, 3.0], dtype='float64')
Index(['1', '2', '3'], dtype='object')
Int64Index([1, 2, 3], dtype='int64')
Index(['1', '2', '3'], dtype='object')
```

or 

All empty Index objects like in case of int vs object.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None

pandas           : 1.0.5
numpy            : 1.19.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 49.1.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
None

</details>

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [ ] closes #35217
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Generally this looks good to me. 

But can you check how well Float64Index `equals` is covered in the tests? 
For example, you added an assert that floats and strings are not equal, but if we preserve behaviour, we should ensure that float / ints and float / object ints keep evaluating to True:

```
In [18]: pd.Index([1, 2]).equals(pd.Index([1., 2.])) 
Out[18]: True

In [20]: pd.Index([1, 2]).equals(pd.Index([1., 2.], dtype=object)) 
Out[20]: True
```

(both of those will work fine as well with the new implementaiton, I think (since array_equivalent right now doesn't check the exact dtype for those, only equality), but just want to ensure it is tested)

**Comment by simonjayhawkins** [comment]
Thanks @jorisvandenbossche for reviewing this. have extended an existing test

File: pandas/tests/indexes/test_numeric.py [hunk 1]
```diff
@@ -232,12 +232,24 @@ def test_equals_numeric(self):
...
-        i = Float64Index([1.0, np.nan])
-        assert i.equals(i)
-        assert i.identical(i)
+        i3 = Float64Index([1.0, np.nan])
+        assert i3.equals(i3)
+        assert i3.identical(i3)
 
-        i2 = Float64Index([1.0, np.nan])
-        assert i.equals(i2)
+        i4 = Float64Index([1.0, np.nan])
```
Comment (jreback) [inline pandas/tests/indexes/test_numeric.py:239]: possible to paramaterize?

File: pandas/tests/indexes/test_numeric.py [hunk 2]
```diff
@@ -635,3 +647,27 @@ def test_uint_index_does_not_convert_to_float64():
     tm.assert_index_equal(result.index, expected)
 
     tm.assert_equal(result, series[:3])
+
+
+def test_float64_index_equals():
+    # https://github.com/pandas-dev/pandas/issues/35217
+    float_index = pd.Index([1.0, 2, 3])
```
Comment (jreback) [inline pandas/tests/indexes/test_numeric.py:654]: ideally parameterize

File: pandas/tests/indexes/test_numeric.py [hunk 3]
```diff
@@ -635,3 +647,27 @@ def test_uint_index_does_not_convert_to_float64():
...
+    result = float_index.equals(string_index)
+    assert result is False
+
+    result = string_index.equals(float_index)
+    assert result is False
+
+
+def test_float64_index_difference():
+    # https://github.com/pandas-dev/pandas/issues/35217
+    float_index = pd.Index([1.0, 2, 3])
```
Comment (jreback) [inline pandas/tests/indexes/test_numeric.py:666]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
@simonjayhawkins are you able to update here?

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins are you able to update here?

i've parameterised the additional tests added in response to https://github.com/pandas-dev/pandas/pull/35231#pullrequestreview-447153446

not sure parameterise makes sense for the issue regression tests (parameterise on `order` i.e index.equals(other) and other.equals(index) or include the indexes in the parameterisation?) 

have pushed changes but not yet appearing here.

**Comment by simonjayhawkins** [comment]
> have pushed changes but not yet appearing here.

The changes are now showing, but ci not running.

**Comment by simonjayhawkins** [comment]
azure and travis triggered, but not github actions. will push another empty commit later if github actions doesn't run.

**Review by TomAugspurger — APPROVED** [review]
All green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. can merge on green.

**Pull request merged**